[Shape Detection article] Better describe feature detection (#4374)
GoogleChrome/web.dev: src/site/content/en/blog/shape-detection/index.md

  - thomassteiner
description: The Shape Detection API detects faces, barcodes, and text in images.
date: 2019-01-07

> updated: 2020-12-11

Author: ```suggestion
updated: 2020-12-14
```

Purely checking for the existence of the constructors to feature detect the

> Shape Detection API doesn't suffice, as Chrome on Linux and Chrome OS

Author: ```suggestion
Shape Detection API doesn't suffice.
```

expose the detectors, but they are known to not work
(which is working as intended according to this [bug](https://crbug.com/920961)).

> Instead, we recommend a *defensive programming* approach by doing feature detection

Reviewer: I don't think this is as clear as it could be. What do you think of this: 

>Purely checking for the existence of the constructors to feature detect the Shape Detection API doesn't suffice.
The presence of an interface doesn't tell you whether the underlying platform supports the feature. This is [as intended](https://crbug.com/920961). It's why we recommend...

Author: Thanks for the suggestion. This is it.

> expose the detectors, but they are known to not work
> (which is working as intended according to this [bug](https://crbug.com/920961)).
> Instead, we recommend a *defensive programming* approach by doing feature detection

Author: ```suggestion
The presence of an interface doesn't tell you whether the underlying platform supports the feature.
This is working [as intended](https://crbug.com/920961).
It's why we recommend a *defensive programming* approach by doing feature detection
```

    .catch(e => e.name === 'NotSupportedError' ? false : true))();
```

The `BarcodeDetector` interface has been updated to include a `getSupportedFormats()` method
and similar interfaces have been proposed
[for `FaceDetector`](https://github.com/WICG/shape-detection-api/issues/53)
and
[for `TextDetector`](https://github.com/WICG/shape-detection-api/issues/57).

```js
await BarcodeDetector.getSupportedFormats();
/* On a macOS computer logs
  [
    "aztec",
    "code_128",
    "code_39",
    "code_93",
    "data_matrix",
    "ean_13",
    "ean_8",
    "itf",
    "pdf417",
    "qr_code",
    "upc_e"
  ]
*/
```

> This allows you to detect precisely for the feature you need, for example, QR code scanning:

Reviewer: ```suggestion
This allows you to detect the speciffic feature you need, for example, QR code scanning:
```

```js
if (('BarcodeDetector' in window) &&
    ((await BarcodeDetector.getSupportedFormats()).includes('qr_code'))) {
  console.log('QR code scanning is supported.');
}
```

> This is better than hiding the interfaces because even across platforms capabilities may vary

Reviewer: ```suggestion
This is better than hiding the interfaces because even across platforms, capabilities may vary
```
